### ISSUE TITLE
torch.utils.dlpack.from_dlpack causes conversion of CuPy float64 tensor to float 16

### ISSUE THREAD

**Issue by Santosh-Gupta** [pytorch/pytorch#24213]
## 🐛 Bug

If I pack a float64 or float32 CuPy tensor using to_dlpack and then use it to set the weights of a Pytorch tensor using from_dlpack, the resulting weights are converted to float16.

## To Reproduce

1. Create NuPy float tensor and pack using to_dlpack . 
2. Create and initialize torch.nn.Embedding tensor .
3. Unpack NuPy float tensor and set it to the Pytorch weight data. 

Code

All the code is in this Google Colab notebook.
https://colab.research.google.com/drive/1fiUcxq6tZMFpT-2tdVYL2h14w21GZhGe
For convenience, the important parts are copied below. 

```
import cupy
import torch

from torch.utils.dlpack import to_dlpack
from torch.utils.dlpack import from_dlpack

import torch.nn as nn

sampl = cupy.random.uniform(low=-1, high=1, size=(131072,512))
sampl
```
will show float32/float64 values

```
emb_size = 131072
emb_dimension = 512

u_embeddings = nn.Embedding(emb_size, emb_dimension, sparse=True).cuda()
initrange = 0.5 / emb_dimension
u_embeddings.weight.data.uniform_(-initrange, initrange)
u_embeddings.weight.data
```
will show float32/float64 values

```
u_embeddings.weight.data = torch.utils.dlpack.from_dlpack( sampl.toDlpack() )
u_embeddings.weight.data

```
will show float16 numbers 

```
u_embeddings.weight.data[0][0]
```

Will show float16 numbers

If I pack and unpack back the CuPy tensor, there is no change

```
dx = sampl.toDlpack()
cx2 = cupy.fromDlpack(dx)
print(cx2)
```

Will display float64/float32

For convenience, here is a Google Colab notebook with all this code 
https://colab.research.google.com/drive/1fiUcxq6tZMFpT-2tdVYL2h14w21GZhGe

## Expected behavior

There should be no data type conversion 

## Environment

Environment is Google Colab. Results of given script

Collecting environment information...
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla K80
Nvidia driver version: 410.79
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.2

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] torch==1.1.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.3.0
[conda] Could not collect

**Comment by pietern** [comment]
Thanks for posting the issue.

I'm not able to reproduce using your script. What I see is the cupy ndarray has dtype equal to float64. Then when I load the tensor with `torch.utils.from_dlpack' I get a tensor with dtype `torch.float64` back, which is what I would expect.

Can you confirm the `dtype` of the tensor you get back?

```python
tensor = torch.utils.dlpack.from_dlpack(sampl.toDlpack())                                                                                                                                                                                                           
print(tensor.dtype)  # --> torch.float64
```

**Comment by Santosh-Gupta** [comment]
> Thanks for posting the issue.
> 
> I'm not able to reproduce using your script. What I see is the cupy ndarray has dtype equal to float64. Then when I load the tensor with `torch.utils.from_dlpack` I get a tensor with dtype `torch.float64` back, which is what I would expect.
> 
> Can you confirm the `dtype` of the tensor you get back?
> 
> ```python
> tensor = torch.utils.dlpack.from_dlpack(sampl.toDlpack())
> print(tensor.dtype)  # --> torch.float64
> ```

Yes, that is what I got back. So I thought it was some sort of truncation, then I decided to look at the numbers using

`u_embeddings.weight.data[0][0].detach().cpu().numpy()`

And it contained the full and original float values. So it looks like Pytorch tensors whose weights where set using CuPy tensors via 'to_dlpack` and `from_dlpack` have some sort of different default display option.

Here's how the weight.data look originally when you initialize the Pytorch tensor directly

```
tensor([[-7.6820e-04,  3.8537e-04, -5.1598e-04,  ..., -3.0010e-04,
          5.0732e-04,  9.3754e-04],
        [ 7.9495e-04,  8.1246e-04, -2.2105e-04,  ..., -7.3008e-04,
         -6.5637e-04, -8.1501e-04],
        [ 5.8629e-04, -3.0736e-04, -2.9700e-05,  ..., -9.4532e-04,
          5.5723e-04, -8.7163e-04],
        ...,
        [ 5.6863e-04,  1.7518e-04, -4.7898e-04,  ..., -2.5984e-04,
          9.7562e-04, -3.7443e-04],
        [ 9.9502e-05, -8.8998e-04,  7.4599e-04,  ..., -2.1337e-04,
          4.2208e-04,  6.5185e-04],
        [-2.8455e-04, -1.5584e-04,  1.6773e-04,  ...,  8.5488e-04,
         -9.2844e-04, -6.8068e-04]], device='cuda:0')
```

But here's how weight.data looks like after the weight.data were set to the CuPy numbers

```
Parameter containing:
tensor([[-0.0479, -0.3545, -0.9477,  ...,  0.8330, -0.0886, -0.7917],
        [ 0.7020,  0.9905, -0.8292,  ...,  0.2256, -0.4756,  0.8602],
        [ 0.6307,  0.6994,  0.6715,  ...,  0.7044, -0.7109, -0.9758],
        ...,
        [-0.4983,  0.7308, -0.8599,  ...,  0.5156, -0.6583,  0.3470],
        [-0.2761,  0.6849,  0.9328,  ...,  0.8455,  0.4019, -0.4411],
        [-0.0502, -0.6109, -0.3718,  ..., -0.5754, -0.5960,  0.8506]],
       device='cuda:0', dtype=torch.float64, requires_grad=True)
```

Another thing I just noticed is that the display starts with 'Parameter containing:'.

I tried looking up display options for pytorch tesnors but I couldn't find anything.

**Comment by pietern** [comment]
My guess is this is caused by the dtype difference. The first tensor uses `torch.float32` and the second tensor (converted from cupy) uses `torch.float64`.

**Comment by Santosh-Gupta** [comment]
I tried with a float32 cupy tensor, using `sampl = cupy.random.uniform(low=-1, high=1, size=(131072,512), dtype=cupy.float32)`  and got the same display type at the end. 

For convenience, here is the notebook that I ran. 

https://colab.research.google.com/drive/1x9M4HK1J7PUoSM-BImTy-P8Qmb_XZwXv

Is it safe to say that it is only a change is display type?

**Comment by mruberry** [comment]
Closing this issue due to age and lack of response. I ran the provided script and see the expected result. @Santosh-Gupta, please reopen if you still believe there is a bug here.
